storage/qingstor.go

```diff
@@ -0,0 +1,187 @@
+	resp, err := q.bucket.HeadObject(objectKey, nil)
+	if err != nil {
+		if e, ok := err.(*errors.QingStorError); ok {
+			if e.StatusCode == http.StatusNotFound {
+				return nil, constants.ErrorQsPathNotFound
+			} else if e.StatusCode == http.StatusForbidden {
+				return nil, constants.ErrorQsPathAccessForbidden
+			}
+		}
+		return
+	}
+
+	om = &ObjectMeta{
+		Key:           objectKey,
+		ContentLength: convert.Int64Value(resp.ContentLength),
+		ContentType:   convert.StringValue(resp.ContentType),
+		ETag:          convert.StringValue(resp.ETag),
+		LastModified:  convert.TimeValue(resp.LastModified),
+		StorageClass:  convert.StringValue(resp.XQSStorageClass),
+	}
+	return
+}
+
+// InitiateMultipartUpload will initiate a multipart upload.
+func (q *QingStorObjectStorage) InitiateMultipartUpload(objectKey string) (uploadID string, err error) {
+	resp, err := q.bucket.InitiateMultipartUpload(objectKey, nil)
+	if err != nil {
+		log.Errorf("Object <%s> InitiateMultipartUpload failed [%v]", objectKey, err)
+		return
+	}
+
+	uploadID = *resp.UploadID
+	return
+}
+
+// UploadMultipart will upload a multipart.
+func (q *QingStorObjectStorage) UploadMultipart(
+	objectKey, uploadID string, size int64, partNumber int, md5sum [16]byte, r io.Reader,
+) (err error) {
+	_, err = q.bucket.UploadMultipart(objectKey, &service.UploadMultipartInput{
+		Body:          r,
+		ContentLength: convert.Int64(size),
+		UploadID:      convert.String(uploadID),
+		PartNumber:    convert.Int(partNumber),
+		ContentMD5:    convert.String(hex.EncodeToString(md5sum[:])),
+	})
+	if err != nil {
+		log.Errorf("Object <%s> part <%d> UploadMultipart failed [%v]", objectKey, partNumber, err)
+		return
+	}
+	return
+}
+
+// CompleteMultipartUpload will complete a multipart upload.
+func (q *QingStorObjectStorage) CompleteMultipartUpload(objectKey, uploadID string, totalNumber int) (err error) {
+	parts := make([]*service.ObjectPartType, totalNumber)
+	for i := 0; i < totalNumber; i++ {
+		parts[i] = &service.ObjectPartType{
+			PartNumber: convert.Int(i),
+		}
+	}
+
+	_, err = q.bucket.CompleteMultipartUpload(
+		objectKey, &service.CompleteMultipartUploadInput{
+			UploadID:    convert.String(uploadID),
+			ObjectParts: parts,
+		})
+	if err != nil {
+		log.Errorf("Object <%s> CompleteMultipartUpload failed [%v]", objectKey, err)
+		return err
+	}
+	return nil
+}
+
+// GetObject will get an object.
+func (q *QingStorObjectStorage) GetObject(objectKey string) (r io.Reader, err error) {
+	resp, err := q.bucket.GetObject(objectKey, nil)
+	if err != nil {
+		log.Errorf("Object <%s> GetObject failed [%v]", objectKey, err)
+		return nil, err
+	}
+	return resp.Body, nil
+}
+
+// PutBucket will make a bucket with specific name.
+func (q *QingStorObjectStorage) PutBucket() error {
+	// Request and create bucket
+	_, err := q.bucket.Put()
+	if err != nil {
+		log.Errorf("Make bucket <%s> in zone <%s> failed [%v]",
+			*q.bucket.Properties.BucketName, *q.bucket.Properties.Zone, err)
```

Comment:
As mentioned before, cannot use `SetupBucket`, so `PutBucket` may not work as expected.